### PR TITLE
initial reinforce, needs to be double checked

### PR DESCRIPTION
**Pull request by dbc148**
I added reinforce to the basic bilstm tagger.  I *think* it's what it should be, but do you mind looking at lines 115 to 135 to make sure I have the theory behind reinforce correct?

### CODE REVIEW

File: 16-reinforce/bilstm-tagger.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    samples = [np.random.choice(range(len(x)), p=x) for x in probs]
+
+    #calculate accuracy=reward
+    correct = [sample == tag for sample, tag in zip(samples, tags)]
+    r_i = float(sum(correct))/len(correct)
+    r = dy.constant((1), r_i)
+    # Reward baseline for each word
+    W_bl = dy.parameter(W_bl_p)
+    b_bl = dy.parameter(b_bl_p)
+    r_b = [dy.affine_transform([b_bl, W_bl, x]) for x in word_reps]
```
Comment (neubig) [inline 16-reinforce/bilstm-tagger.py]: I think you actually need to break the computation graph here by wrapping `x` in `dy.nobackprop()`

File: 16-reinforce/bilstm-tagger.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    #we need to take the value in order to break the computation graph
+    #as the reward portion is trained seperatley and not backpropogated through during the overall score
+    rewards_over_baseline = [(r - x).value() for x in r_b]
```
Comment (neubig) [inline 16-reinforce/bilstm-tagger.py]: You can also do `dy.nobackprop()` here as well

**Review by neubig — CHANGES_REQUESTED** [review]
Thanks! This looks pretty close: just two minor changes.

File: 16-reinforce/bilstm-tagger.py (hunk 1 above)
Comment (dbc148) [reply 16-reinforce/bilstm-tagger.py]: Good call!  I was using pretrained word representations in my c++ code and completely missed that here.

File: 16-reinforce/bilstm-tagger.py (hunk 2 above)
Comment (dbc148) [reply 16-reinforce/bilstm-tagger.py]: Shouldn't taking the value do that?  Or is this instead of taking the value?

**Pull request merged**